ldsocache.go

```
@@ -108,6 +139,193 @@ func (shdr *LDSOCacheExtensionSectionHeader) describe() {
+		ver = pieces[len(pieces)-1]
+	}
+
+	fmt.Printf("RETURNING %s for %s\n", name, realname)
+	return name, ver, nil
+}
+
+func AddLDSOCacheEntriesForDir(fsys fs.FS, libdir string, entries *[]LDSOCacheEntry) (error) {
+	// fs.FS wants all file paths to be relative
+	libdir = strings.TrimLeft(libdir, "/")
+	dirents, err := fs.ReadDir(fsys, libdir)
+	if err != nil {
+		// It is OK for a directory to not exist
+		return nil
+	}
+
+	for _, dirent := range dirents {
+		realname := dirent.Name()
+		fullpath := filepath.Join(libdir, realname)
+		mode := dirent.Type()
+		is_link := (mode & fs.ModeSymlink != 0)
+
+		if is_link {
+			// Stat follows symlinks
+			info, err := fs.Stat(fsys, fullpath)
+			if err != nil {
+				fmt.Printf("Warning: Could not stat %s\n", fullpath)
+				continue
+			}
+			if !info.Mode().IsRegular() {
+				fmt.Printf("DEBUG: Skipping %s, not a link to a regular file\n", fullpath)
+				continue
+			}
+		}
+
+		if !(mode.IsRegular() || is_link) {
+			continue
+		}
+		libf, err := fsys.Open(fullpath)
+		if err != nil {
+			continue
+		}
+		defer libf.Close()
+		var libfReaderAt io.ReaderAt
+		libfReaderAt, ok := libf.(io.ReaderAt)
+		if !ok {
+			// Ugly: Work around lack of ReaderAt support by
+			// reading the entire file into memory
+			var buf []byte
+			buf, err = fs.ReadFile(fsys, fullpath)
+			if err != nil {
+				fmt.Printf("DEBUG: Unable to open %s\n", fullpath)
+				continue
+			}
+			libf.Close()
+			libfReaderAt = bytes.NewReader(buf)
+		}
+		elflibf, err := elf.NewFile(libfReaderAt)
+
+		if err != nil {
+			fmt.Printf("DEBUG: Unable to open %s as ELF\n", fullpath)
+			continue
+		}
+		// FIXME: do we need to check for the ELF magic bytes?
+		if elflibf.FileHeader.Type != elf.ET_DYN {
+			continue
+		}
+		flags := uint32(0)
+		flags |= FLAG_ELF
+		// FIXME: Shouldn't just assert this
+		flags |= FLAG_ELF_LIBC6
+		sonames, err := elflibf.DynString(elf.DT_SONAME)
+		if err != nil {
+			continue
```

Is it okay to ignore this error?

---

I think so, because `ldconfig` does:

```
$ sudo touch /usr/lib/x86_64-linux-gnu/libspider.so.9
$ ldconfig -C /tmp/cache 
/sbin/ldconfig.real: File /lib/x86_64-linux-gnu/libspider.so.9 is empty, not checked.
$ echo notanelf | sudo tee /usr/lib/x86_64-linux-gnu/libspider.so.9
notanelf
$ ldconfig -C /tmp/cache 
$ echo $?
0
$ ldconfig -p -N -C /tmp/cache | grep spider
$
```